File: Units/LibRawMxWrapper.pas
{$IFDEF FPC} {$MODE DELPHI} {$ENDIF}

unit LibRawMxWrapper;

interface

uses
  Windows, SysUtils;

const LibRawWrapper = {$IFDEF WIN64}'LibRawMxWrapper_s_crt_64.dll'{$ELSE}'LibRawMxWrapper_s_crt.dll'{$ENDIF};

var
  RawProcessorCreate: function : Pointer; stdcall;
  RawProcessorFree: procedure (RawProcessor: Pointer); stdcall;
  RawProcessorStrError: function (Reserved: Pointer; P: Integer): PChar; stdcall;
  RawProcessorVersion: function : PChar; stdcall;
  RawProcessorOpenFile: function (RawProcessor: Pointer; FileName: PChar): Integer; stdcall;
  RawProcessorSizes: procedure (
    RawProcessor: Pointer;
    var RawWidth, RawHeight: Word;
    var Width, Height: Word;
    var LeftMargin, TopMargin: Word;
    var Iwidth, Iheight: Word;
    var RawPitch: LongWord;
    var PixelAspect: Double;
    var Flip: Integer); stdcall;
  RawProcessorAdjustSizesInfoOnly: function (RawProcessor: Pointer): Integer; stdcall;
  RawProcessorUnpack: function (RawProcessor: Pointer): Integer; stdcall;
  RawProcessorCheck: function (RawProcessor: Pointer): Integer; stdcall;
  RawProcessorRawImage: function (RawProcessor: Pointer): PWord; stdcall;
  RawProcessorMake: function (RawProcessor: Pointer): PChar; stdcall;
  RawProcessorModel: function (RawProcessor: Pointer): PChar; stdcall;
  RawProcessorSoftware: function (RawProcessor: Pointer): PChar; stdcall;
  RawProcessorTimestamp: function (RawProcessor: Pointer): Int64; stdcall;
  RawProcessorTime: procedure (RawProcessor: Pointer; TimeStr: PChar; TimeStrLen: Integer); stdcall;
  RawProcessorShutter: function (RawProcessor: Pointer): Single; stdcall;
  RawProcessorISOspeed: function (RawProcessor: Pointer): Single; stdcall;
  RawProcessorBayerPattern: procedure (RawProcessor: Pointer; BayerPatternStr: PChar; BayerPatternStrLen: Integer); stdcall;

procedure InitLibrary(const DLLname: string);

implementation

function InitFunctionPointer(H: THandle; const ProcName: string; const DLLname: string): Pointer;
begin
  Result := GetProcAddress(H, PChar(ProcName));
  if (Result = nil) then
    raise Exception.Create('Cannot find procedure ' + ProcName + ' in ' + DLLname);
end;

procedure InitLibrary(const DLLname: string);
var
  H: THandle;
begin
  H := SafeLoadLibrary(DLLname);
  if H = 0 then RaiseLastOSError;
{$IFDEF WIN64}
  RawProcessorCreate := InitFunctionPointer(H, 'RawProcessorCreate', DLLname);
  RawProcessorFree := InitFunctionPointer(H, 'RawProcessorFree', DLLname);
  RawProcessorStrError := InitFunctionPointer(H, 'RawProcessorStrError', DLLname);
  RawProcessorVersion := InitFunctionPointer(H, 'RawProcessorVersion', DLLname);
  RawProcessorOpenFile := InitFunctionPointer(H, 'RawProcessorOpenFile', DLLname);
  RawProcessorSizes := InitFunctionPointer(H, 'RawProcessorSizes', DLLname);
  RawProcessorAdjustSizesInfoOnly := InitFunctionPointer(H, 'RawProcessorAdjustSizesInfoOnly', DLLname);
  RawProcessorUnpack := InitFunctionPointer(H, 'RawProcessorUnpack', DLLname);
  RawProcessorCheck := InitFunctionPointer(H, 'RawProcessorCheck', DLLname);
  RawProcessorRawImage := InitFunctionPointer(H, 'RawProcessorRawImage', DLLname);
  RawProcessorMake := InitFunctionPointer(H, 'RawProcessorMake', DLLname);
  RawProcessorModel := InitFunctionPointer(H, 'RawProcessorModel', DLLname);
  RawProcessorSoftware := InitFunctionPointer(H, 'RawProcessorSoftware', DLLname);
  RawProcessorTimestamp := InitFunctionPointer(H, 'RawProcessorTimestamp', DLLname);
  RawProcessorTime := InitFunctionPointer(H, 'RawProcessorTime', DLLname);
  RawProcessorShutter := InitFunctionPointer(H, 'RawProcessorShutter', DLLname);
  RawProcessorISOspeed := InitFunctionPointer(H, 'RawProcessorISOspeed', DLLname);
  RawProcessorBayerPattern := InitFunctionPointer(H, 'RawProcessorBayerPattern', DLLname);
{$ELSE}
  RawProcessorCreate := InitFunctionPointer(H, '_RawProcessorCreate@0', DLLname);
  RawProcessorFree := InitFunctionPointer(H, '_RawProcessorFree@4', DLLname);
  RawProcessorStrError := InitFunctionPointer(H, '_RawProcessorStrError@8', DLLname);
  RawProcessorVersion := InitFunctionPointer(H, '_RawProcessorVersion@0', DLLname);
  RawProcessorOpenFile := InitFunctionPointer(H, '_RawProcessorOpenFile@8', DLLname);
  RawProcessorSizes := InitFunctionPointer(H, '_RawProcessorSizes@48', DLLname);
  RawProcessorAdjustSizesInfoOnly := InitFunctionPointer(H, '_RawProcessorAdjustSizesInfoOnly@4', DLLname);
  RawProcessorUnpack := InitFunctionPointer(H, '_RawProcessorUnpack@4', DLLname);
  RawProcessorCheck := InitFunctionPointer(H, '_RawProcessorCheck@4', DLLname);
  RawProcessorRawImage := InitFunctionPointer(H, '_RawProcessorRawImage@4', DLLname);
  RawProcessorMake := InitFunctionPointer(H, '_RawProcessorMake@4', DLLname);
  RawProcessorModel := InitFunctionPointer(H, '_RawProcessorModel@4', DLLname);
  RawProcessorSoftware := InitFunctionPointer(H, '_RawProcessorSoftware@4', DLLname);
  RawProcessorTimestamp := InitFunctionPointer(H, '_RawProcessorTimestamp@4', DLLname);
  RawProcessorTime := InitFunctionPointer(H, '_RawProcessorTime@12', DLLname);
  RawProcessorShutter := InitFunctionPointer(H, '_RawProcessorShutter@4', DLLname);
  RawProcessorISOspeed := InitFunctionPointer(H, '_RawProcessorISOspeed@4', DLLname);
  RawProcessorBayerPattern := InitFunctionPointer(H, '_RawProcessorBayerPattern@12', DLLname);
{$ENDIF}
end;

initialization
  RawProcessorCreate := nil;
  RawProcessorFree := nil;
  RawProcessorStrError := nil;
  RawProcessorVersion := nil;
  RawProcessorOpenFile := nil;
  RawProcessorSizes := nil;
  RawProcessorAdjustSizesInfoOnly := nil;
  RawProcessorUnpack := nil;
  RawProcessorCheck := nil;
  RawProcessorRawImage := nil;
  RawProcessorMake := nil;
  RawProcessorModel := nil;
  RawProcessorSoftware := nil;
  RawProcessorTimestamp := nil;
  RawProcessorTime := nil;
  RawProcessorShutter := nil;
  RawProcessorISOspeed := nil;
  RawProcessorBayerPattern := nil;
end.

File: LibRawWrapper/LibRawMxWrapper/LibRawMxWrapper.cpp
// LibRawMxWrapper.cpp
//

#include "stdafx.h"

#define S RawProcessor->imgdata.sizes
#define I RawProcessor->imgdata.idata
#define R RawProcessor->imgdata.rawdata
#define O RawProcessor->imgdata.other

__declspec(dllexport) const char* __stdcall RawProcessorVersion(void)
{
	return LibRaw::version();
}

__declspec(dllexport) LibRaw* __stdcall RawProcessorCreate(void)
{
	return new LibRaw;
}

__declspec(dllexport) void __stdcall RawProcessorFree(LibRaw* RawProcessor)
{
	delete RawProcessor;
}

__declspec(dllexport) const char* __stdcall RawProcessorStrError(void* reserved, int p)
{
	return LibRaw::strerror(p);
}

__declspec(dllexport) int __stdcall RawProcessorOpenFile(LibRaw* RawProcessor, char* filename)
{
	return RawProcessor->open_file(filename);
}

__declspec(dllexport) void __stdcall RawProcessorSizes(
	LibRaw* RawProcessor,
	ushort &raw_width, ushort &raw_height,
	ushort &width, ushort &height,
	ushort &left_margin, ushort &top_margin, 
	ushort &iwidth, ushort &iheight,
	unsigned &raw_pitch,
	double &pixel_aspect,
	int &flip
)
{
	raw_width = S.raw_width;
	raw_height = S.raw_height;
	width = S.width;
	height = S.height;
	left_margin = S.left_margin;
	top_margin = S.top_margin;
	iwidth = S.iwidth;
	iheight = S.iheight;
	raw_pitch = S.raw_pitch;
	pixel_aspect = S.pixel_aspect;
	flip = S.flip;
}

__declspec(dllexport) int __stdcall RawProcessorAdjustSizesInfoOnly(LibRaw* RawProcessor)
{
	return RawProcessor->adjust_sizes_info_only();
}

__declspec(dllexport) int __stdcall RawProcessorUnpack(LibRaw* RawProcessor)
{
	return RawProcessor->unpack();
}

__declspec(dllexport) int __stdcall RawProcessorCheck(LibRaw* RawProcessor)
{
	if (!(I.filters || I.colors == 1)) return -1; else return 0;
}

__declspec(dllexport) ushort* __stdcall RawProcessorRawImage(LibRaw* RawProcessor)
{

	return R.raw_image;
}

__declspec(dllexport) time_t __stdcall RawProcessorTimestamp(LibRaw* RawProcessor)
{
	return O.timestamp;
}

__declspec(dllexport) void __stdcall RawProcessorTime(LibRaw* RawProcessor, char *timestr, int timest_len)
{
	time_t t = O.timestamp;
	ctime_s(timestr, timest_len, &t);
}

__declspec(dllexport) float __stdcall RawProcessorShutter(LibRaw* RawProcessor)
{
	return O.shutter;
}

__declspec(dllexport) float __stdcall RawProcessorISOspeed(LibRaw* RawProcessor)
{
	return O.iso_speed;
}

__declspec(dllexport) char* __stdcall RawProcessorMake(LibRaw* RawProcessor)
{
	return R.iparams.make;
}

__declspec(dllexport) char* __stdcall RawProcessorModel(LibRaw* RawProcessor)
{
	return R.iparams.model;
}

__declspec(dllexport) char* __stdcall RawProcessorSoftware(LibRaw* RawProcessor)
{
	return R.iparams.software;
}

__declspec(dllexport) void __stdcall RawProcessorBayerPattern(LibRaw* RawProcessor, char *bayer_pattern, int bayer_pattern_len)
{
	char color_desc[5];
	for (int i = 0; i < 5; i++) color_desc[i] = I.cdesc[i];
	//putchar('\n'); for (int i = 0; i < 5; i++) { putchar('"'); putchar(color_desc[i]); putchar('"'); putchar('\n'); }
	if (color_desc[3]) color_desc[3] = 'G';
    for (int i = 0; i < 16 || i < bayer_pattern_len - 1; i++)
	{
		bayer_pattern[i] = color_desc[RawProcessor->fcol(i >> 1, i & 1)];
		bayer_pattern[i+1] = '\0';
	}

}

#undef O
#undef R
#undef I
#undef S
